# Projekt IAU  
## 1. časť - Prieskumná analýza 
### Vypracovali: Martin Kukučka, Martin Jankuliak

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import scipy.stats as stats
import glob

import datetime
from datetime import datetime,date
from datetime import timedelta
from dateutil.relativedelta import relativedelta

import json
import re
import dateutil.parser as parser
import statsmodels.api as sm

import scipy as sc
from pandas import read_csv


### Načítanie dát

In [3]:
df_other_train = read_csv("data/other_train.csv")
df_personal_train = read_csv("data/personal_train.csv")

## Základný opis dát spolu s ich charakteristikami 

### Dataset other_train.csv

Pre predstavu tu je ukážka datasetu

In [4]:
df_other_train.head()

Unnamed: 0             name  \
0           0    Robert Plante   
1           1     Richard Huey   
2           2     Joseph Sable   
3           3  Gregory Preston   
4           4       Mike Riley   

                                             address                 race  \
0  3594 Johnson Green Suite 277\r\nJohnsonport, V...                White   
1             350 Lewis Haven\r\nAngelaton, IN 39486   Asian-Pac-Islander   
2  22509 Adrian Canyon Suite 799\r\nEast Timothyt...                White   
3           2674 Daryl Island\r\nLake Sean, TN 89363                White   
4  023 Joseph Estate Suite 799\r\nLake Andrewvill...                White   

        marital-status          occupation pregnant  education-num  \
0              Widowed       Other-service        f            5.0   
1        Never-married   Handlers-cleaners        f           13.0   
2        Never-married     Exec-managerial        f            9.0   
3   Married-civ-spouse     Exec-managerial        f           13.0   
4   Married-civ-spouse   Handlers-cleaners        f            9.0   

     relationship  capital-gain   education    fnlwgt  class  income  \
0       Unmarried           0.0         9th  220019.0    1.0   <=50K   
1   Not-in-family           0.0   Bachelors  196899.0    0.0   <=50K   
2       Unmarried           0.0     HS-grad  301591.0    0.0   <=50K   
3         Husband           0.0   Bachelors  310320.0    1.0    >50K   
4         Husband           0.0     HS-grad  237943.0    0.0   <=50K   

                                        medical_info  native-country  \
0  {'mean_glucose':'106.71875','std_glucose':'48....   United-States   
1  {'mean_glucose':'118.609375','std_glucose':'na...           Haiti   
2  {'mean_glucose':'129.4609375','std_glucose':'5...   United-States   
3  {'mean_glucose':'30.1015625','std_glucose':'37...   United-States   
4  {'mean_glucose':'124.0390625','std_glucose':'4...   United-States   

   hours-per-week  capital-loss workclass  
0             9.0           0.0   Private  
1            50.0           0.0   Private  
2            35.0           0.0   Private  
3            40.0           0.0   Private  
4            60.0           0.0   private

Pomocou tohto príkazu zistíme počet riadkov

In [5]:
df_other_train.shape[0]

3983

A teraz počet atribútov pre tento dataset

In [6]:
df_other_train.shape[1]

19

Typy atribútov zistíme týmto príkazom

In [7]:
df_other_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3983 entries, 0 to 3982
Data columns (total 19 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Unnamed: 0      3983 non-null   int64  
 1   name            3983 non-null   object 
 2   address         3983 non-null   object 
 3   race            3691 non-null   object 
 4   marital-status  3970 non-null   object 
 5   occupation      3961 non-null   object 
 6   pregnant        3969 non-null   object 
 7   education-num   3961 non-null   float64
 8   relationship    3954 non-null   object 
 9   capital-gain    3969 non-null   float64
 10  education       3972 non-null   object 
 11  fnlwgt          3968 non-null   float64
 12  class           3960 non-null   float64
 13  income          3972 non-null   object 
 14  medical_info    3966 non-null   object 
 15  native-country  3966 non-null   object 
 16  hours-per-week  3966 non-null   float64
 17  capital-loss    3973 non-null   f

Zistili sme, že tento dataset je zložený z atribútov:
- 6x float
- 1x int
- 12x object

Ďalej tu vidíme aj počet riadkov: 3983 ("RangeIndex: 3983 entries"),
a počet stĺpcov: 19 ("Data columns (total 19 columns)")

Nakoniec sa pozrieme na desktiptívne štatistiky.

In [8]:
df_other_train.describe()

Unnamed: 0  education-num  capital-gain         fnlwgt        class  \
count  3983.00000    3961.000000   3969.000000    3968.000000  3960.000000   
mean   1991.00000      10.052764   1134.530108  189963.996976     0.256313   
std    1149.93739       2.520019   8067.028170  102327.438172     0.436652   
min       0.00000       1.000000      0.000000   20308.000000     0.000000   
25%     995.50000       9.000000      0.000000  119475.250000     0.000000   
50%    1991.00000      10.000000      0.000000  180585.500000     0.000000   
75%    2986.50000      12.000000      0.000000  236173.000000     1.000000   
max    3982.00000      16.000000  99999.000000  795830.000000     1.000000   

       hours-per-week  capital-loss  
count     3966.000000   3973.000000  
mean        40.203984     72.926252  
std         12.317838    360.910859  
min          1.000000      0.000000  
25%         40.000000      0.000000  
50%         40.000000      0.000000  
75%         45.000000      0.000000  
max         99.000000   2824.000000

V tejto tabuľke vidíme štatistiky ako sú: počet záznamov (nenulových), priemer, štandardnú odchýlku, minimum, prvý kvartil, druhý kvartil (medián), tretí kvartil, maximum, pre všetky atribúty, ktoré obsahujú numerické hodnoty.

### Dataset personal_train.csv

Rovnako vykonáme opis dát aj pre druhý dataset

Ukážka datasetu

In [9]:
df_personal_train.head()

Unnamed: 0             name  \
0           0  Roscoe Bohannon   
1           1     Ernest Kline   
2           2  Harold Hendriks   
3           3   Randy Baptiste   
4           4  Anthony Colucci   

                                             address  age      sex  \
0  7183 Osborne Ways Apt. 651\r\nEast Andrew, OH ...   58     Male   
1  391 Ball Road Suite 961\r\nFlowersborough, IN ...   43   Female   
2     8702 Vincent Square\r\nNew Jerryfurt, CO 30614   51     Male   
3     2751 Harris Crossroad\r\nWest Ashley, CA 30311   66     Male   
4  904 Robert Cliffs Suite 186\r\nWest Kyle, CO 7...   45   Female   

  date_of_birth  
0    1961-04-27  
1    1976-09-12  
2    1967-11-12  
3    1953-06-02  
4    1974/09/11

Počet riadkov

In [10]:
df_personal_train.shape[0]

3933

Počet atribútov

In [11]:
df_personal_train.shape[1]

6

Typy atribútov

In [12]:
df_personal_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3933 entries, 0 to 3932
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     3933 non-null   int64 
 1   name           3933 non-null   object
 2   address        3933 non-null   object
 3   age            3933 non-null   int64 
 4   sex            3933 non-null   object
 5   date_of_birth  3933 non-null   object
dtypes: int64(2), object(4)
memory usage: 184.5+ KB


Zistile sme, že tento dataset je zložený z atribútov:
- 2x int
- 4x object

Ďalej tu vidíme aj počet riadkov: 3933 ("RangeIndex: 3933 entries"),
a počet stĺpcov: 6 ("Data columns (total 6 columns)")

A Nakoniec desktiptívne štatistiky.

In [13]:
df_personal_train.describe()

Unnamed: 0          age
count  3933.000000  3933.000000
mean   1966.000000    51.816934
std    1135.503633    11.814803
min       0.000000    -1.000000
25%     983.000000    44.000000
50%    1966.000000    52.000000
75%    2949.000000    60.000000
max    3932.000000   113.000000

## Párová analýza dát

## Identifikácia problémov v dátach spolu s predpokladaným scenárom riešenia v ďalšej fáze

Nevhodná štruktúra dát sa nachádza v atribúte medical_info v datasete other_train.csv, kde je atribút zložený z viacero zložiek (mean_glucose, std_glucose, kurtosis_glucose, skewness_glucose, mean_oxygen, std_oxygen, kurtosis_oxygen, skewness_oxygen), tento atribút chceme rozdeliť do viacero stĺpcov. Taktiež aj atribút address obsahuje viac zložiek, ale nepredpokladáme, že by hodnota Y významne závisela na tomto atribúte, preto počítame s tým, že tento atribút nebudeme podrobnejšie deliť.

In [24]:
df_other_train.shape[0] - df_personal_train.shape[0]

50

Zistili sme, že dataset other_train.csv obsahuje obsahuje o 50 záznamov viac ako personal_train. Predpokladáme, že v datasete other_train.csv sa budú nachádzať duplicitné záznamy.

In [14]:
df_other_train.duplicated().sum()

0

Duplicitné riadky sme nanašli, ale ešte nás zaujíma atribút name, pretože predpokladáme, že každý participant ma unikátne meno a nemáme viac záznamov daného človeka.

In [15]:
df_other_train.name.duplicated().sum()

50

Našli sme 50 duplicít, čo zodpovedá rozdielu záznamov medzi týmito datasetmi, vypíšeme si pár z nich.

In [25]:
df_other_train[df_other_train.name.isin(df_other_train.name[df_other_train.name.duplicated()])].sort_values(by=['name']).name.head(6)

2078    Alex Stubblefield
2682    Alex Stubblefield
2416         Ashley Lyons
467          Ashley Lyons
2553          Billy Clark
3408          Billy Clark
Name: name, dtype: object

Rovnaký postup zvolíme aj pre druhý dataset

In [17]:
df_personal_train.duplicated().sum()

0

In [18]:
df_personal_train.name.duplicated().sum()

0

Zistili sme, že v druhom datasete sa žiadne duplicity nenachádzajú.

Duplicitné záznamy chceme nahradiť len 1 záznamom a v prípade chýbajúcich hodnôt doplníme tieto hodnoty z druhého záznamu, ak budú existovať.

Teraz ideme nájsť nejednotné formáty dát. Budú nás zaujímať atribúty pregnant, workclass a z druhého datasetu age a date_of_birth.

In [73]:
df_other_train['pregnant'].unique()

array(['f', 'FALSE', 'F', 't', nan, 'TRUE', 'T'], dtype=object)

In [74]:
df_other_train['workclass'].unique()

array([' Private', 'private', ' Local-gov', ' State-gov', nan,
       ' Self-emp-inc', ' Self-emp-not-inc', 'state-gov', ' Federal-gov',
       ' ?', 'self-emp-not-inc', 'local-gov', 'federal-gov', '?',
       'self-emp-inc', ' Never-worked', ' Without-pay'], dtype=object)

In [75]:
age_formats = df_personal_train['age'].astype(str).str.replace('\d', '0')
age_formats.unique()

array(['00', '0', '000', '-0'], dtype=object)

In [76]:
date_formats = df_personal_train.date_of_birth
date_formats = date_formats.str.replace('\d', '0')
date_formats.unique()

array(['0000-00-00', '0000/00/00', '0000-00-00 00:00:00',
       '0000-00-00 00 00 00', '00/00/0000', '00-00-00'], dtype=object)

Nejednotné formáty dát sme našli. Formáty predstavujúce rovnakú hodnotu chceme zjednotiť na jeden tvar. V atribúte age vidíme záporné aj trojciferné hodnoty, tieto hodnoty zaradíme medzi vychýlené hodnoty a následne k ním vypočítame správnú hodnotu pomocou atribútu date_of_birth.

In [44]:
df_other_train.isnull().sum()

Unnamed: 0          0
name                0
address             0
race              292
marital-status     13
occupation         22
pregnant           14
education-num      22
relationship       29
capital-gain       14
education          11
fnlwgt             15
class              23
income             11
medical_info       17
native-country     17
hours-per-week     17
capital-loss       10
workclass          17
dtype: int64

Ako vidíme, chýbajú nám hodnoty v atribútoch 3-18, tieto hodnoty chceme doplniť mediánom alebo priemerom.

In [45]:
df_personal_train.isnull().sum()

Unnamed: 0       0
name             0
address          0
age              0
sex              0
date_of_birth    0
dtype: int64

V druhom datasete nechýbajú žiadne hodnoty.

Vychýlené hodnoty plánujeme nahradiť mediánom alebo priemerom.

## 2. časť - Predspracovanie

Spájanie dát

In [125]:
df_other_valid = read_csv("data/other_valid.csv")
df_personal_valid = read_csv("data/personal_valid.csv")

df_merged1 = pd.merge(df_personal_train, df_other_train, on=['name', 'address'])
df_merged2 = pd.merge(df_personal_valid, df_other_valid, on=['name', 'address'])

df_merged2

Unnamed: 0_x             name  \
0                0    Philip Miller   
1                1     Mitch Wilson   
2                2      James Olsen   
3                3    Maurice Riley   
4                4    Larry Stanley   
...            ...              ...   
1356          1306  Michael Everett   
1357          1307   Julian Patrick   
1358          1308       Lee Walker   
1359          1309  Thomas Garrison   
1360          1310       Brian Mann   

                                                address  age      sex  \
0     7910 Rosales Plain Apt. 454\r\nPort Carl, GA 6...   28     Male   
1     055 Morgan Plains Suite 225\r\nEast Darrylmout...   60     Male   
2                    PSC 7359, Box 2088\r\nAPO AE 62717   81     Male   
3                             USS Allen\r\nFPO AP 87677   61     Male   
4       491 Fields Key Suite 544\r\nGracestad, MT 13456   43   Female   
...                                                 ...  ...      ...   
1356  35892 Robert Forest Apt. 556\r\nRubioside, AL ...   53     Male   
1357        53981 Taylor Brook\r\nBarkerville, TN 17001   32     Male   
1358  847 Jamie Cove Apt. 464\r\nCummingsview, RI 90999   39     Male   
1359  159 Yvonne Way Suite 696\r\nWest Ashleyton, LA...   32   Female   
1360  05550 Mahoney Crest Apt. 037\r\nGrahammouth, M...   51     Male   

     date_of_birth  Unnamed: 0_y    race       marital-status  \
0       1991-02-22           693   White   Married-civ-spouse   
1       1959/06/15           215   White   Married-civ-spouse   
2       24/01/1938          1182   White   Married-civ-spouse   
3       1958-05-02          1169   Black   Married-civ-spouse   
4       1976-08-16           166   White             Divorced   
...            ...           ...     ...                  ...   
1356    1966-01-17           670   White        Never-married   
1357    1987-09-10          1012   White   Married-civ-spouse   
1358    1979-11-30           299   White        Never-married   
1359    1987-10-08           371   Black        Never-married   
1360    1968-02-15            43   White   Married-civ-spouse   

              occupation  ... capital-gain      education    fnlwgt  class  \
0      Machine-op-inspct  ...          0.0   Some-college  176186.0    0.0   
1         Prof-specialty  ...          0.0      Bachelors  402998.0    0.0   
2           Adm-clerical  ...          0.0        HS-grad  304651.0    1.0   
3        Exec-managerial  ...      20051.0    Prof-school  162945.0    0.0   
4           Adm-clerical  ...          0.0   Some-college   41809.0    1.0   
...                  ...  ...          ...            ...       ...    ...   
1356   Handlers-cleaners  ...          0.0        HS-grad  243178.0    0.0   
1357     Exec-managerial  ...          0.0      Bachelors  178249.0    0.0   
1358               Sales  ...          0.0      Bachelors  416338.0    0.0   
1359   Machine-op-inspct  ...          0.0      Assoc-voc  221324.0    0.0   
1360        Craft-repair  ...          0.0   Some-college  308136.0    1.0   

      income                                       medical_info  \
0       >50K  {'mean_glucose':'147.5','std_glucose':'-4323.3...   
1       >50K  {'mean_glucose':'131.25','std_glucose':'47.327...   
2       >50K  {'mean_glucose':'109.0546875','std_glucose':'4...   
3       >50K  {'mean_glucose':'123.8515625','std_glucose':'n...   
4      <=50K  {'mean_glucose':'68.125','std_glucose':'35.699...   
...      ...                                                ...   
1356   <=50K  {'mean_glucose':'118.3046875','std_glucose':'4...   
1357    >50K  {'mean_glucose':'116.5703125','std_glucose':'4...   
1358   <=50K  {'mean_glucose':'97.7421875','std_glucose':'41...   
1359   <=50K  {'mean_glucose':'73.4921875','std_glucose':'-4...   
1360   <=50K  {'mean_glucose':'111.890625','std_glucose':'47...   

      native-country hours-per-week capital-loss   workclass  
0      United-States           40.0          0.0     Private  
1      Unit

In [124]:
dir_in = 'data'
filenames = [i for i in glob.glob(dir_in + '/*')]
print(*filenames, sep='\n')

data\other_train.csv
data\other_valid.csv
data\personal_train.csv
data\personal_valid.csv
